### PR TITLE
Provide a way of cancelling worker node shutdown

### PR DESCRIPTION
**Pull request by samarthjain**
Cancelling shutdown comes in handy in environments where we dynamically
scale up or down the presto worker nodes based on the load.

Consider the case when there a cluster is being scaled down because there
isn't enough load. Now, when new demand comes in, due to capacity constraint,
we may need to wait for scaling down activity to complete before we can add
new capacity to the cluster. By having a way of cancelling worked node shutdown,
we can quickly scale back up by not having to wait for active tasks to complete
or by not having to terminate the JVM..

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/server/GracefulShutdownHandler.java [hunk 1]
```diff
@@ -89,8 +91,11 @@ public synchronized void requestShutdown()
             List<TaskInfo> activeTasks = getActiveTasks();
             while (activeTasks.size() > 0) {
                 CountDownLatch countDownLatch = new CountDownLatch(activeTasks.size());
-
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/server/GracefulShutdownHandler.java]: unnecessary change.

**Comment by samarthjain** [comment]
Thanks for the review, @nezihyigitbasi . I got rid of the SHUTTING_DOWN state and now allow a node to transition from shutting_down to active. I also refactored the test to better test out the shutdown and cancel shutdown flow. 

One thing I wanted to bring your attention on was the minor race condition here - https://github.com/prestodb/presto/blob/dfcbc6a0b941aae33d6113179a148ec8c610a3e0/presto-main/src/main/java/com/facebook/presto/server/ServerInfoResource.java#L84. 

I can go ahead and make the operation thread safe (by locking on shutdownHandler), if you would like. I left it as is since the race condition shouldn't have much of an impact.

Thanks!

**Comment by samarthjain** [comment]
@nezihyigitbasi - pinging to see if you got a chance to look at the updated PR. Thanks!

**Comment by samarthjain** [comment]
@nezihyigitbasi - would you have cycles now to review this change? I have cleaned up the code a bit, added new tests and addressed the comments you made in your last review. Thanks!

File: presto-main/src/main/java/com/facebook/presto/server/ServerInfoResource.java [hunk 2]
```diff
@@ -100,11 +113,13 @@ public Response updateState(NodeState state)
     @Produces(APPLICATION_JSON)
     public NodeState getServerState()
     {
-        if (shutdownHandler.isShutdownRequested()) {
-            return SHUTTING_DOWN;
-        }
-        else {
-            return ACTIVE;
+        synchronized (shutdownHandler) {
+            if (shutdownHandler.isShutdownRequested()) {
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/server/ServerInfoResource.java]: `isShutdownRequested` is already synchronized, why do you need the synch. block?
